### PR TITLE
Fixed for block griefing

### PR DESCRIPTION
**Pull request by liach**
See CovertJaguar/Railcraft-API#15

### CODE REVIEW

**Comment by liach** [comment]
Reference #280

File: src/main/java/mods/railcraft/common/plugins/forge/PlayerPlugin.java [hunk 1]
```diff
@@ -42,12 +69,15 @@ public static void writeOwnerToNBT(NBTTagCompound nbt, GameProfile owner) {
...
     public static GameProfile readOwnerFromNBT(NBTTagCompound nbt) {
-        String ownerName = "[Unknown]";
+        String ownerName = null;
         if (nbt.hasKey("owner"))
             ownerName = nbt.getString("owner");
         UUID ownerUUID = null;
         if (nbt.hasKey("ownerId"))
             ownerUUID = UUID.fromString(nbt.getString("ownerId"));
+        if (ownerName == null || ownerUUID == null) {
+            return CartTools.RAILCRAFT_PROFILE;
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/plugins/forge/PlayerPlugin.java:79]: No...just no...

File: src/main/java/mods/railcraft/common/blocks/machine/gamma/TileDispenserCart.java [hunk 2]
```diff
@@ -125,7 +127,7 @@ public void onPulse() {
                             float rx = MiscTools.getRand().nextFloat() * 0.8F + 0.1F;
                             float ry = MiscTools.getRand().nextFloat() * 0.8F + 0.1F;
                             float rz = MiscTools.getRand().nextFloat() * 0.8F + 0.1F;
-                            EntityItem item = new EntityItem(worldObj, x + rx, y + ry, z + rz, cartStack);
+                            Entity item = cartStack.getItem().createEntity(worldObj, new EntityItem(worldObj, x, y, z, cartStack), cartStack);
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/blocks/machine/gamma/TileDispenserCart.java]: I really need to track this function down...

File: src/main/java/mods/railcraft/common/carts/EntityCartUndercutter.java [hunk 3]
```diff
@@ -174,6 +180,19 @@ private void replaceWith(int x, int y, int z, int slotExist, int slotStock) {
...
+            }
+            Block placedAgainst = WorldPlugin.getBlock(worldObj, x, y - 1, z);
+            BlockEvent.PlaceEvent placeEvent = new BlockEvent.PlaceEvent(
+                    new BlockSnapshot(worldObj, x, y, z, stockBlock, 0), placedAgainst,
+                    PlayerPlugin.getFakePlayer((WorldServer) worldObj, x, y, z, this)
+            );
+            MinecraftForge.EVENT_BUS.post(placeEvent);
+            if (placeEvent.isCanceled()) {
+                return;
+            }
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/carts/EntityCartUndercutter.java:195]: Can we move this into WorldPlugin? Maybe as a new function, not sure.

File: src/main/java/mods/railcraft/common/plugins/forge/PlayerPlugin.java [hunk 4]
```diff
@@ -70,7 +100,7 @@ public static String getUsername(World world, UUID playerId) {
             if (player != null)
                 return player.getDisplayName();
         }
-        return "[Unknown]";
+        return "[Railcraft]";
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/plugins/forge/PlayerPlugin.java:103]: This probably breaks stuff...I remember some "[Unknown]".equals(string) somewhere.

File: src/main/java/mods/railcraft/common/plugins/forge/PlayerPlugin.java [hunk 5]
```diff
@@ -61,7 +91,7 @@ public static String getUsername(World world, GameProfile gameProfile) {
         String username = gameProfile.getName();
         if (username != null && !username.equals(""))
             return username;
-        return "[Unknown]";
+        return "[Railcraft]";
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/plugins/forge/PlayerPlugin.java:94]: This probably breaks stuff...I remember some "[Unknown]".equals(string) somewhere.

**Review by CovertJaguar — CHANGES_REQUESTED** [review]

File: src/main/java/mods/railcraft/common/blocks/machine/gamma/TileDispenserCart.java (hunk 2 above)
Comment (CovertJaguar) [reply src/main/java/mods/railcraft/common/blocks/machine/gamma/TileDispenserCart.java]: Yes, this is wrong, see #905

**Comment by liach** [comment]
Well I will try to push some change tomorrow.

File: src/main/java/mods/railcraft/common/plugins/forge/PlayerPlugin.java [hunk 6]
```diff
@@ -20,16 +22,41 @@
...
 
+import java.util.UUID;
+
 /**
  *
  * @author CovertJaguar <http://www.railcraft.info/>
  */
 public class PlayerPlugin {
 
-    public static final GameProfile RAILCRAFT_USER_PROFILE = new GameProfile(UUID.nameUUIDFromBytes("[Railcraft]".getBytes()), "[Railcraft]");
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/plugins/forge/PlayerPlugin.java:29]: This stays.

**Review by CovertJaguar — CHANGES_REQUESTED** [review]

File: src/main/java/mods/railcraft/common/plugins/forge/PlayerPlugin.java (hunk 5 above)
Comment (CovertJaguar) [reply src/main/java/mods/railcraft/common/plugins/forge/PlayerPlugin.java:94]: Also... "[Railcraft]" should generally have similar to admin permissions, whereas "[Unknown]" would be fully restricted. So returning "[Railcraft]" as a fallback is not generally safe.

File: src/main/java/mods/railcraft/common/blocks/machine/gamma/TileDispenserCart.java (hunk 2 above)
Comment (liach) [inline src/main/java/mods/railcraft/common/blocks/machine/gamma/TileDispenserCart.java]: So will you try to merge this pull when I fixed this patch?

Comment (liach) [reply src/main/java/mods/railcraft/common/blocks/machine/gamma/TileDispenserCart.java]: Fixed

**Comment by liach** [comment]
Should I open a pull which I pick these fixes to 1.10 branch?

**Comment by CovertJaguar** [comment]
@liach yeah, something like that

**Comment by liach** [comment]
Should close this one

**Pull request closed without merging**